active/0004-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,405 @@
+         "properties": {
+            "extensionConfigs": {
+               "k8s": {
+                  "namespace": {
+                     "type": "string",
+                     "value": "default"
+                  },
+                  "kubeConfig": {
+                     "type": "directive",
+                     "evaluation": [
+                        {
+                           "type": "ArmApiCall",
+                           "resourceId": "/subscriptions/.../resourceGroups/.../Microsoft.ContainerService/managedClusters/...",
+                           "method": "GET",
+                           "apiVersion": "2024-02-01",
+                           "apiAction": "/listClusterAdminCredential",
+                           "query": "?a=1&b=2" // optional
+                           "body": {} // optional
+                        },
+                        {
+                           "type": "JSONPath",
+                           "path": "kubeconfig[0].value"
+                        }
+                     ]
+                  }
+               }
+            },
+            "template": {
+               "extensions": {
+                  "k8s": {
+                     "extension": "Kubernetes",
+                     "version": "1.0.0"
+                     // "config" is removed, it's expected to be provided by properties of the deployment
+                  }
+               }
+            }
+         }
+      }
+   }
+}
+```
+
+### Microsoft.Resources/deployments API changes
+
+This REP extends upon the API changes in REP 0003 by adding a configuration value to the extensions returned from
+a deployment GET. This configuration object will be the same as the configuration object described in the above ARM
+template design changes.
+
+```json
+{
+   // ... deployment contract
+   "extensions": [
+      {
+         "name": "Kubernetes",
+         "alias": "k8s",
+         "version": "1.0.0",
+         "deploymentId": "/subscriptions/.../resourceGroups/.../providers/Microsoft.Resources/deployments/...",
+         "config": {
+            "namespace": {
+               "type": "string",
+               "value": "default"
+            },
+            "kubeConfig": {
+               "type": "directive",
+               "evaluation": [
+                  {
+                     "type": "ArmApiCall",
+                     "resourceId": "/subscriptions/.../resourceGroups/.../Microsoft.ContainerService/managedClusters/...",
+                     "method": "GET",
+                     "apiVersion": "2024-02-01",
+                     "apiAction": "/listClusterAdminCredential",
+                     "query": "?a=1&b=2" // optional
+                     "body": { } // optional
+                  },
+                  {
+                     "type": "JSONPath",
+                     "path": "kubeconfig[0].value" // decompose?
+                  }
+               ]
+            }
+         }
+      }
+   ]
+}
+```
```

Comment:
I like the example with creating an AKS cluster and then passing in the kubecontext to a child module.

What's the "right" way to do this for a stack when the extension config is passed at the top level? 

- Imagine my AKS cluster already exists, do we need additional support for the CLI to work with extension config?
- Imagine it's not an AKS cluster 😆 should I store the config in keyvault?


Your

Reply:
1. One way might be to extend the bicepparam/json file with an extensions section. It would allow referencing ARM resources but prohibit other sources (inlining, loadFileContent, env vars, etc) for secure portion of the configuration schema. Other bits like `namespace` could be expressed like parameters are.

2. Yes, this will be supported. That specific example is not outlined here yet. I think the general constraint to start with is that expressions are restricted to ARM resource accesses expressions. I detailed this some more in recent commits to the Bicep example.

Reply:
Added design proposals for bicepparam/JSON parameter file changes.